File: src/124. Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */

export function maxPathSum(root) {
  let max = -Infinity
  dfs(root)
  return max
  function dfs(node) {
    if (!node) return -Infinity
    const leftSum = dfs(node.left)
    const rightSum = dfs(node.right)
    const currentSum = node.val + leftSum + rightSum
    const emitSum = node.val + Math.max(leftSum, rightSum, 0)
    max = Math.max(max, leftSum, rightSum, currentSum, emitSum)
    return emitSum
  }
}

File: src/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans = -1000;
    int dfs(TreeNode* root) {
        if (!root) return -1000;
        int leftSum = dfs(root->left),
            rightSum = dfs(root->right),
            twoPathSum = root->val + leftSum + rightSum,
            singlePathSum = root->val + max({ leftSum, rightSum, 0 });
        ans = max({ ans, leftSum, rightSum, singlePathSum, twoPathSum });
        return singlePathSum;
    }
public:
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return ans;
    }
};